dali/operators/generic/cast.h

```diff
@@ -17,45 +17,62 @@
 
 #include <vector>
 
-#include "dali/pipeline/operator/operator.h"
 #include "dali/core/convert.h"
+#include "dali/core/dev_buffer.h"
+#include "dali/core/tensor_shape.h"
+#include "dali/kernels/common/block_setup.h"
+#include "dali/pipeline/operator/operator.h"
 
 namespace dali {
 
+struct CastSampleDesc {
+  void *output;
+  const void *input;
+};
 
-#define CAST_ALLOWED_TYPES \
-  (bool, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t, \
-  float16, float, double)
+#define CAST_ALLOWED_TYPES                                                                         \
+  (bool, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t, float16, float, \
+  double)
 
 template <typename Backend>
 class Cast : public Operator<Backend> {
  public:
-  explicit inline Cast(const OpSpec &spec) :
-    Operator<Backend>(spec),
-    output_type_(spec.GetArgument<DALIDataType>("dtype"))
-    {}
+  explicit inline Cast(const OpSpec &spec)
+      : Operator<Backend>(spec), output_type_(spec.GetArgument<DALIDataType>("dtype")) {}
 
   inline ~Cast() override = default;
 
   DISABLE_COPY_MOVE_ASSIGN(Cast);
 
  protected:
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
   bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<Backend> &ws) override {
-    return false;
+    output_desc.resize(1);
+    const auto &input = ws.template InputRef<Backend>(0);
+    output_desc[0].shape = input.shape();
+    output_desc[0].type = TypeTable::GetTypeInfo(output_type_);
+    PrepareBlocks(ws);
+    return true;
   }
 
-  void RunImpl(Workspace<Backend> &ws) override;
+  void PrepareBlocks(const workspace_t<Backend> &ws);
 
- private:
-  template <typename OType, typename IType>
-  inline void CPUHelper(OType *out, const IType *in, size_t N) {
-    for (size_t i = 0; i < N; ++i) {
-      out[i] = ConvertSat<OType>(in[i]);
-    }
-  }
+  void RunImpl(workspace_t<Backend> &ws) override;
 
+ private:
   DALIDataType output_type_;
 
+  using GpuBlockSetup = kernels::BlockSetup<1, -1>;
+
+  GpuBlockSetup block_setup_;
+  std::vector<CastSampleDesc> samples_;
+  DeviceBuffer<GpuBlockSetup::BlockDesc> blocks_dev_;
+  DeviceBuffer<CastSampleDesc> samples_dev_;
```

Comment:
Rather nitpick, but with this we have `GpuBlockSetup` also for CPU operator. Is it ok?

Reply:
Shouldn't be a big of an issue if it's unused. I don't want to do a super big rework.